File: backend/drivers/javascript/driver.js
const fs = require('fs');
const solution = require('./solution');

const className = process.env.ALGOVERSE_CLASS_NAME;
const funcName = process.env.ALGOVERSE_FUNCTION_NAME;
let paramNames = [];
try {
    const parsedParamNames = JSON.parse(process.env.ALGOVERSE_PARAM_NAMES || '[]');
    if (Array.isArray(parsedParamNames)) {
        paramNames = parsedParamNames.filter((name) => typeof name === 'string' && name.length > 0);
    }
} catch (error) {
    paramNames = [];
}
if (!funcName) {
    console.error('ALGOVERSE_FUNCTION_NAME environment variable not set');
    process.exit(1);
}

let targetFunc;
let targetInstance = null;

if (className && solution[className]) {
    targetInstance = new solution[className]();
    targetFunc = targetInstance[funcName];
} else {
    targetFunc = solution[funcName];
}

if (typeof targetFunc !== 'function') {
    console.error(`Function '${funcName}' not found in solution.js`);
    process.exit(1);
}

const stringifyOutput = (value) => {
    if (value === undefined) return '';
    try {
        const serialized = JSON.stringify(value);
        return typeof serialized === 'string' ? serialized : String(value);
    } catch (error) {
        return String(value);
    }
};

const stringifyArg = (value) => {
    if (typeof value === 'string') return value;
    try {
        return JSON.stringify(value);
    } catch (error) {
        return String(value);
    }
};

(async () => {
    const inputRaw = fs.readFileSync(0, 'utf-8');
    let batchInputs = [];

    try {
        batchInputs = JSON.parse(inputRaw);
    } catch (error) {
        batchInputs = [];
    }

    if (!Array.isArray(batchInputs)) batchInputs = [];

    const results = [];

    for (const input of batchInputs) {
        const start = process.hrtime.bigint();
        const startHeap = process.memoryUsage().heapUsed;
        let stdout = '';
        let stderr = '';
        let printedOutput = '';
        let returnMissing = false;

        const capturedLogs = [];
        const originalConsole = {
            log: console.log,
            info: console.info,
            warn: console.warn,
            error: console.error
        };

        const captureLog = (...args) => {
            capturedLogs.push(args.map(stringifyArg).join(' '));
        };

        console.log = captureLog;
        console.info = captureLog;
        console.warn = captureLog;
        console.error = captureLog;

        try {
            let parsedArgs = input;
            try {
                parsedArgs = JSON.parse(input);
            } catch (error) {
                // Keep raw string input.
            }

            let out;
            if (Array.isArray(parsedArgs)) {
                out = targetFunc.apply(targetInstance, parsedArgs);
            } else if (parsedArgs && typeof parsedArgs === 'object') {
                const orderedArgs = paramNames.length > 0
                    ? paramNames.map((name) => parsedArgs[name])
                    : Object.values(parsedArgs);
                out = targetFunc.apply(targetInstance, orderedArgs);
            } else {
                out = targetFunc.call(targetInstance, parsedArgs);
            }
            if (out && typeof out.then === 'function') {
                out = await out;
            }

            returnMissing = out === undefined;
            stdout = returnMissing ? '' : stringifyOutput(out);
            printedOutput = capturedLogs.join('\n').trim();

            if (printedOutput) {
                stderr = 'Use return statement for final answer. Printed output is shown below.';
            } else if (returnMissing) {
                stderr = 'Function must return output using return statement.';
            }
        } catch (error) {
            printedOutput = capturedLogs.join('\n').trim();
            stderr = error && error.message ? error.message : String(error);
            returnMissing = false;
        } finally {
            console.log = originalConsole.log;
            console.info = originalConsole.info;
            console.warn = originalConsole.warn;
            console.error = originalConsole.error;
        }

        const end = process.hrtime.bigint();
        const runtime = Number(end - start) / 1e6;
        const endHeap = process.memoryUsage().heapUsed;
        const memoryMb = Math.max(startHeap, endHeap) / (1024 * 1024);

        results.push({
            stdout,
            stderr,
            printedOutput,
            returnMissing,
            time: runtime.toFixed(3),
            memory: Number(memoryMb.toFixed(3))
        });
    }

    process.stdout.write(JSON.stringify(results));
})().catch((error) => {
    console.error(`Driver Fatal Error: ${error.message || String(error)}`);
    process.exit(1);
});

File: backend/drivers/python/driver.py
import sys
import os
import io
import json
import time
import contextlib
import tracemalloc

# Ensure generated stdout/stderr are captured correctly
sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')
sys.stdin = io.TextIOWrapper(sys.stdin.buffer, encoding='utf-8')

import solution  # type: ignore[import-not-found]

func_name = os.environ.get('ALGOVERSE_FUNCTION_NAME')
class_name = os.environ.get('ALGOVERSE_CLASS_NAME')

if not func_name:
    sys.stderr.write("ALGOVERSE_FUNCTION_NAME environment variable not set\n")
    sys.exit(1)

# Retrieve the function or class method
if class_name and hasattr(solution, class_name):
    cls = getattr(solution, class_name)
    instance = cls()
    if not hasattr(instance, func_name):
        sys.stderr.write(f"Function '{func_name}' not found in class '{class_name}'\n")
        sys.exit(1)
    func = getattr(instance, func_name)
else:
    if not hasattr(solution, func_name):
        sys.stderr.write(f"Function '{func_name}' not found in solution.py\n")
        sys.exit(1)
    func = getattr(solution, func_name)

try:
    # Read entire stdin as JSON Batch
    input_data = sys.stdin.read()
    batch_inputs = json.loads(input_data)
    
    if not isinstance(batch_inputs, list):
        raise ValueError("Input must be a JSON array")

    results = []

    for inp in batch_inputs:
        tracemalloc.start()
        tracemalloc.reset_peak()
        start_time = time.time()
        memory_mb: float = 0.0
        try:
            # Simple heuristic: try JSON parse, else raw str
            arg = inp
            try:
                arg = json.loads(inp)
            except:
                pass

            captured_stdout = io.StringIO()
            with contextlib.redirect_stdout(captured_stdout):
                if isinstance(arg, dict):
                    # Unpack JSON object as keyword arguments
                    ret = func(**arg)
                elif isinstance(arg, list):
                    # Unpack JSON array as positional arguments
                    ret = func(*arg)
                else:
                    # Single argument
                    ret = func(arg)
            
            printed_output = captured_stdout.getvalue().strip()
            return_missing = ret is None

            if return_missing:
                out = ""
            else:
                try:
                    out = json.dumps(ret, ensure_ascii=False)
                except Exception:
                    out = json.dumps(str(ret), ensure_ascii=False)

            duration = (time.time() - start_time) * 1000 # ms
            _current, peak = tracemalloc.get_traced_memory()
            memory_mb = peak / (1024 * 1024)
            error_message = ""
            if printed_output:
                error_message = "Use return statement for final answer. Printed output is shown below."
            elif return_missing:
                error_message = "Function must return output using return statement."

            results.append({
                "stdout": out,
                "stderr": error_message,
                "printedOutput": printed_output,
                "returnMissing": return_missing,
                "time": duration,
                "memory": int(memory_mb * 1000) / 1000
            })
        except Exception as e:
            results.append({
                "stdout": "",
                "stderr": str(e),
                "printedOutput": "",
                "returnMissing": False,
                "time": 0,
                "memory": 0
            })
        finally:
            tracemalloc.stop()

    print(json.dumps(results))

except Exception as e:
    sys.stderr.write(f"Driver Error: {str(e)}\n")
    sys.exit(1)
